crates/llm-ls/src/adaptors.rs

```diff
@@ -0,0 +1,163 @@
+use super::{
+    internal_error, APIError, APIResponse, CompletionParams, Generation, Ide, NAME, VERSION,
+};
+use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, USER_AGENT};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tower_lsp::jsonrpc;
+
+fn build_tgi_body(prompt: String, params: CompletionParams) -> Value {
+    serde_json::json!({
+        "inputs": prompt,
+        "parameters": params.request_params,
+    })
+}
+
+fn build_tgi_headers(api_token: Option<&String>, ide: Ide) -> Result<HeaderMap, jsonrpc::Error> {
+    let mut headers = HeaderMap::new();
+    let user_agent = format!("{NAME}/{VERSION}; rust/unknown; ide/{ide:?}");
+    headers.insert(
+        USER_AGENT,
+        HeaderValue::from_str(&user_agent).map_err(internal_error)?,
+    );
+
+    if let Some(api_token) = api_token {
+        headers.insert(
+            AUTHORIZATION,
+            HeaderValue::from_str(&format!("Bearer {api_token}")).map_err(internal_error)?,
+        );
+    }
+
+    Ok(headers)
+}
+
+fn parse_tgi_text(text: reqwest::Result<String>) -> Result<Vec<Generation>, jsonrpc::Error> {
+    let generations =
+        match serde_json::from_str(&text.unwrap_or_default()).map_err(internal_error)? {
+            APIResponse::Generation(gen) => vec![gen],
+            APIResponse::Generations(_) => {
+                return Err(internal_error(
+                    "TGI parser unexpectedly encountered api-inference",
+                ))
+            }
+            APIResponse::Error(err) => return Err(internal_error(err)),
+        };
+    Ok(generations)
+}
+
+fn build_api_body(prompt: String, params: CompletionParams) -> Value {
+    build_tgi_body(prompt, params)
+}
+
+fn build_api_headers(api_token: Option<&String>, ide: Ide) -> Result<HeaderMap, jsonrpc::Error> {
+    build_tgi_headers(api_token, ide)
+}
+
+fn parse_api_text(text: reqwest::Result<String>) -> Result<Vec<Generation>, jsonrpc::Error> {
+    let generations =
+        match serde_json::from_str(&text.unwrap_or_default()).map_err(internal_error)? {
+            APIResponse::Generation(gen) => vec![gen],
+            APIResponse::Generations(gens) => gens,
+            APIResponse::Error(err) => return Err(internal_error(err)),
+        };
+    Ok(generations)
+}
+
+fn build_ollama_body(prompt: String, params: CompletionParams) -> Value {
```

Comment:
```suggestion
fn build_ollama_body(prompt: String, params: &CompletionParams) -> Value {
```
and all the others as well, afaict we don't need to pass the value, the ref is enough